server/src/main/java/org/elasticsearch/search/query/QueryPhaseCollector.java

```diff
@@ -143,31 +144,36 @@ public LeafCollector getLeafCollector(LeafReaderContext context) throws IOExcept
         Bits postFilterBits = getPostFilterBits(context);
 
         if (aggsCollector == null) {
-            LeafCollector topDocsLeafCollector;
+            LeafCollector tdlc = null;
             try {
-                topDocsLeafCollector = topDocsCollector.getLeafCollector(context);
+                tdlc = topDocsCollector.getLeafCollector(context);
             } catch (@SuppressWarnings("unused") CollectionTerminatedException e) {
                 // TODO we keep on collecting although we have nothing to collect (there is no top docs nor aggs leaf collector).
                 // The reason is only to set the early terminated flag to the QueryResult like some tests expect. This needs fixing.
                 if (terminateAfter == 0) {
                     throw e;
                 }
-                topDocsLeafCollector = null;
+            }
+            final LeafCollector topDocsLeafCollector = tdlc;
+            if (postFilterWeight == null && terminateAfter == 0 && minScore == null) {
+                // no need to wrap if we just need to collect unfiltered docs through leaf collector
+                // aggs collector was not originally provided so the overall score mode is that of the top docs collector
+                return topDocsLeafCollector;
             }
```

Comment:
Could we instead check that `postFilterBits == null`? It seems like that is the key thing. We don't need to apply the post filter bits by wrapping. 

If that is not acceptable, then we should have an `assert postFilterBits == null;` in this early exit case.

Reply:
that's ok. I believe that postFilterBits are going to be null only when postFilterWeight is null as well, or is there something I am not seing?

Reply:
>  I believe that postFilterBits are going to be null only when postFilterWeight is null as well, or is there something I am not seing?

This is true, but its in another method. We have so many predicts and exits from this function I am just being paranoid.

Reply:
makes sense, thanks for raising. Maybe the inner leaf collector classes should be static so they don't have direct access to all the state of the top-level collector class. I think this is an improvement that I may be able to make as a follow-up, I did not do it to start with because I needed to carry around a lot of context anyways.